SignalServiceKit/src/Messages/OWSMessageManager.m

```diff
@@ -1053,6 +1070,70 @@ - (TSIncomingMessage *_Nullable)handleReceivedEnvelope:(OWSSignalServiceProtosEn
     }
 }
 
+- (TSQuotedMessage *_Nullable)quotedMessageForDataMessage:(OWSSignalServiceProtosDataMessage *)dataMessage
+{
+    OWSAssert(dataMessage);
+
+    if (!dataMessage.hasQuote) {
+        return nil;
+    }
+
+    OWSSignalServiceProtosDataMessageQuote *quoteProto = [dataMessage quote];
+
+    if (![quoteProto hasId] || [quoteProto id] == 0) {
+        OWSFail(@"%@ quoted message missing id", self.logTag);
+        return nil;
+    }
+    uint64_t timestamp = [quoteProto id];
+
+    if (![quoteProto hasAuthor] || [quoteProto author].length == 0) {
+        OWSFail(@"%@ quoted message missing author", self.logTag);
+        return nil;
+    }
+    // TODO: We could verify that this is a valid e164 value.
+    NSString *authorId = [quoteProto author];
+
+    NSString *_Nullable body = nil;
+    BOOL hasText = NO;
+    BOOL hasAttachment = NO;
+    if ([quoteProto hasText] && [quoteProto text].length > 0) {
+        body = [quoteProto text];
+        hasText = YES;
+    }
+
+    NSString *_Nullable sourceFilename = nil;
+    NSData *_Nullable thumbnailData = nil;
+    NSString *_Nullable contentType = nil;
+
+    if ([quoteProto hasAttachment]) {
+        OWSSignalServiceProtosAttachmentPointer *attachmentProto = [quoteProto attachment];
+        if ([attachmentProto hasContentType] && attachmentProto.contentType.length > 0) {
+            contentType = attachmentProto.contentType;
+
+            if ([attachmentProto hasFileName] && attachmentProto.fileName.length > 0) {
+                sourceFilename = attachmentProto.fileName;
+            }
+            if ([attachmentProto hasThumbnail] && attachmentProto.thumbnail.length > 0) {
+                thumbnailData = [attachmentProto thumbnail];
+            }
+        }
+        hasAttachment = YES;
+    }
+
+    if (!hasText && !hasAttachment) {
+        OWSFail(@"%@ quoted message has neither text nor attachment", self.logTag);
+        return nil;
+    }
+
+    TSQuotedMessage *quotedMessage = [[TSQuotedMessage alloc] initWithTimestamp:timestamp
```

Comment:
`timestamp` is the timestamp of the message that we're quoting (e.g. **not** the quotedMessage.timestamp)

Maybe this should be like:

```
uint64_t quotingTimestamp = [quoteProto id];
...
[[TSQuotedMessage alloc] initQuotedMessageWithTimestamp:[NSDate ows_millisecondTimestamp]
                                       quotingTimestamp:quotingTimestamp
...
```

Reply:
So `TSQuotedMessage` represents the data of _just_ the quoted message, not the "quoting reply".  The quoting replies are ordinary instances of `TSIncomingMessage` and `TSOutgoingMessage` whose `quotedMessage` property is set.